File: packages/client/ios/RNOta/Store/FileIo.swift
import Foundation

public enum AtomicWriteCrashPoint: Sendable {
  case none
  case afterTempWrite
  case afterTempFsync
  case beforeRename
  case afterRename
}

public struct SimulatedCrashError: Error, Equatable {
  public let message: String
  public init(_ message: String) { self.message = message }
}

/// Filesystem boundary for the store (injectable for crash / disk-full tests).
public protocol FileIo: Sendable {
  func writeAtomic(target: URL, temporary: URL, bytes: Data) throws
  func readData(from file: URL) throws -> Data
  func fileExists(_ url: URL) -> Bool
  func isFile(_ url: URL) -> Bool
  func isDirectory(_ url: URL) -> Bool
  func createDirectory(_ url: URL) throws
  func removeItem(_ url: URL) throws
  func listDirectoryNames(_ url: URL) throws -> [String]
  func moveItem(from: URL, to: URL) throws
}

public struct RealFileIo: FileIo {
  public init() {}

  public func writeAtomic(target: URL, temporary: URL, bytes: Data) throws {
    do {
      try FileManager.default.createDirectory(
        at: temporary.deletingLastPathComponent(),
        withIntermediateDirectories: true
      )
      try bytes.write(to: temporary, options: .atomic)
      let handle = try FileHandle(forWritingTo: temporary)
      defer { try? handle.close() }
      try handle.synchronize()
      try moveItem(from: temporary, to: target)
    } catch let error as BundleStoreException {
      throw error
    } catch {
      if Self.isNoSpace(error) {
        try? FileManager.default.removeItem(at: temporary)
        throw BundleStoreException.diskFull
      }
      throw BundleStoreException.ioFailure("Failed to atomically write \(target.lastPathComponent)")
    }
  }

  public func readData(from file: URL) throws -> Data {
    try Data(contentsOf: file)
  }

  public func fileExists(_ url: URL) -> Bool {
    FileManager.default.fileExists(atPath: url.path)
  }

  public func isFile(_ url: URL) -> Bool {
    var isDir: ObjCBool = false
    guard FileManager.default.fileExists(atPath: url.path, isDirectory: &isDir) else { return false }
    return !isDir.boolValue
  }

  public func isDirectory(_ url: URL) -> Bool {
    var isDir: ObjCBool = false
    guard FileManager.default.fileExists(atPath: url.path, isDirectory: &isDir) else { return false }
    return isDir.boolValue
  }

  public func createDirectory(_ url: URL) throws {
    try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
  }

  public func removeItem(_ url: URL) throws {
    if fileExists(url) {
      try FileManager.default.removeItem(at: url)
    }
  }

  public func listDirectoryNames(_ url: URL) throws -> [String] {
    guard isDirectory(url) else { return [] }
    return try FileManager.default.contentsOfDirectory(atPath: url.path)
  }

  public func moveItem(from: URL, to: URL) throws {
    let fm = FileManager.default
    if fm.fileExists(atPath: to.path) {
      try fm.removeItem(at: to)
    }
    try fm.moveItem(at: from, to: to)
  }

  private static func isNoSpace(_ error: Error) -> Bool {
    let ns = error as NSError
    return ns.domain == NSPOSIXErrorDomain && ns.code == Int(ENOSPC)
  }
}

/// Test double that can simulate mid-write crashes and disk-full errors.
public final class ControllableFileIo: FileIo, @unchecked Sendable {
  private let delegate: FileIo
  public var crashPoint: AtomicWriteCrashPoint
  public var failWriteWithDiskFull: Bool

  public init(
    delegate: FileIo = RealFileIo(),
    crashPoint: AtomicWriteCrashPoint = .none,
    failWriteWithDiskFull: Bool = false
  ) {
    self.delegate = delegate
    self.crashPoint = crashPoint
    self.failWriteWithDiskFull = failWriteWithDiskFull
  }

  public func writeAtomic(target: URL, temporary: URL, bytes: Data) throws {
    if failWriteWithDiskFull {
      throw BundleStoreException.diskFull
    }
    try FileManager.default.createDirectory(
      at: temporary.deletingLastPathComponent(),
      withIntermediateDirectories: true
    )
    switch crashPoint {
    case .none:
      try delegate.writeAtomic(target: target, temporary: temporary, bytes: bytes)
    case .afterTempWrite:
      try bytes.write(to: temporary, options: .atomic)
      throw SimulatedCrashError("crash after temp write")
    case .afterTempFsync:
      try bytes.write(to: temporary, options: .atomic)
      let handle = try FileHandle(forWritingTo: temporary)
      try handle.synchronize()
      try handle.close()
      throw SimulatedCrashError("crash after temp fsync")
    case .beforeRename:
      try bytes.write(to: temporary, options: .atomic)
      let handle = try FileHandle(forWritingTo: temporary)
      try handle.synchronize()
      try handle.close()
      throw SimulatedCrashError("crash before rename")
    case .afterRename:
      try bytes.write(to: temporary, options: .atomic)
      let handle = try FileHandle(forWritingTo: temporary)
      try handle.synchronize()
      try handle.close()
      try delegate.moveItem(from: temporary, to: target)
      throw SimulatedCrashError("crash after rename")
    }
  }

  public func readData(from file: URL) throws -> Data { try delegate.readData(from: file) }
  public func fileExists(_ url: URL) -> Bool { delegate.fileExists(url) }
  public func isFile(_ url: URL) -> Bool { delegate.isFile(url) }
  public func isDirectory(_ url: URL) -> Bool { delegate.isDirectory(url) }
  public func createDirectory(_ url: URL) throws { try delegate.createDirectory(url) }
  public func removeItem(_ url: URL) throws { try delegate.removeItem(url) }
  public func listDirectoryNames(_ url: URL) throws -> [String] { try delegate.listDirectoryNames(url) }
  public func moveItem(from: URL, to: URL) throws { try delegate.moveItem(from: from, to: to) }
}

File: packages/client/android/src/main/java/com/rnota/store/FileIo.kt
package com.rnota.store

import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption

/**
 * Filesystem boundary for the store.
 * Production uses [DEFAULT]; unit tests inject crash / disk-full behavior.
 */
interface FileIo {
  fun writeAtomic(
    target: File,
    temporary: File,
    bytes: ByteArray,
  )

  fun readBytes(file: File): ByteArray

  fun exists(file: File): Boolean

  fun isFile(file: File): Boolean

  fun isDirectory(file: File): Boolean

  fun mkdir(dir: File): Boolean

  fun deleteRecursively(file: File): Boolean

  fun listNames(dir: File): List<String>

  fun rename(from: File, to: File): Boolean

  companion object {
    val DEFAULT: FileIo = RealFileIo()
  }
}

/**
 * Crash-injection points for atomic state writes (test harness).
 */
enum class AtomicWriteCrashPoint {
  NONE,
  AFTER_TEMP_WRITE,
  AFTER_TEMP_FSYNC,
  BEFORE_RENAME,
  AFTER_RENAME,
}

internal class RealFileIo : FileIo {
  override fun writeAtomic(
    target: File,
    temporary: File,
    bytes: ByteArray,
  ) {
    try {
      temporary.parentFile?.mkdirs()
      FileOutputStream(temporary).use { fos ->
        fos.write(bytes)
        fos.fd.sync()
      }
      atomicMove(temporary, target)
      syncParent(target)
    } catch (e: IOException) {
      if (isNoSpace(e)) {
        temporary.delete()
        throw BundleStoreException.DiskFull(e)
      }
      throw BundleStoreException.IoFailure("Failed to atomically write ${target.name}", e)
    }
  }

  override fun readBytes(file: File): ByteArray = file.readBytes()

  override fun exists(file: File): Boolean = file.exists()

  override fun isFile(file: File): Boolean = file.isFile

  override fun isDirectory(file: File): Boolean = file.isDirectory

  override fun mkdir(dir: File): Boolean = dir.mkdirs() || dir.isDirectory

  override fun deleteRecursively(file: File): Boolean = file.deleteRecursively()

  override fun listNames(dir: File): List<String> = dir.list()?.toList().orEmpty()

  override fun rename(
    from: File,
    to: File,
  ): Boolean = atomicMove(from, to)

  private fun atomicMove(
    from: File,
    to: File,
  ): Boolean {
    try {
      Files.move(
        from.toPath(),
        to.toPath(),
        StandardCopyOption.REPLACE_EXISTING,
        StandardCopyOption.ATOMIC_MOVE,
      )
      return true
    } catch (_: IOException) {
      // Some platforms reject ATOMIC_MOVE; fall back to replace.
      Files.move(from.toPath(), to.toPath(), StandardCopyOption.REPLACE_EXISTING)
      return true
    }
  }

  private fun syncParent(file: File) {
    val parent = file.parentFile ?: return
    try {
      FileOutputStream(parent).use { it.fd.sync() }
    } catch (_: Exception) {
      // Directory fsync is best-effort; not available on all Android versions.
    }
  }

  private fun isNoSpace(e: IOException): Boolean {
    val message = e.message?.lowercase().orEmpty()
    return message.contains("no space") ||
      message.contains("enospc") ||
      e.javaClass.simpleName.contains("ENOSPC", ignoreCase = true)
  }
}

/**
 * Test double that can simulate mid-write crashes and disk-full errors.
 */
class ControllableFileIo(
  private val delegate: FileIo = FileIo.DEFAULT,
  var crashPoint: AtomicWriteCrashPoint = AtomicWriteCrashPoint.NONE,
  var failWriteWithDiskFull: Boolean = false,
) : FileIo by delegate {
  override fun writeAtomic(
    target: File,
    temporary: File,
    bytes: ByteArray,
  ) {
    if (failWriteWithDiskFull) {
      throw BundleStoreException.DiskFull()
    }
    temporary.parentFile?.mkdirs()
    when (crashPoint) {
      AtomicWriteCrashPoint.NONE -> delegate.writeAtomic(target, temporary, bytes)
      AtomicWriteCrashPoint.AFTER_TEMP_WRITE -> {
        temporary.writeBytes(bytes)
        throw SimulatedCrashException("crash after temp write")
      }
      AtomicWriteCrashPoint.AFTER_TEMP_FSYNC -> {
        FileOutputStream(temporary).use { fos ->
          fos.write(bytes)
          fos.fd.sync()
        }
        throw SimulatedCrashException("crash after temp fsync")
      }
      AtomicWriteCrashPoint.BEFORE_RENAME -> {
        FileOutputStream(temporary).use { fos ->
          fos.write(bytes)
          fos.fd.sync()
        }
        throw SimulatedCrashException("crash before rename")
      }
      AtomicWriteCrashPoint.AFTER_RENAME -> {
        FileOutputStream(temporary).use { fos ->
          fos.write(bytes)
          fos.fd.sync()
        }
        delegate.rename(temporary, target)
        throw SimulatedCrashException("crash after rename")
      }
    }
  }
}

class SimulatedCrashException(
  message: String,
) : RuntimeException(message)
